qp_deblur/deblur.py

```diff
@@ -73,20 +74,69 @@ def generate_deblur_workflow_commands(preprocessed_fp, out_dir, parameters):
     return cmd
 
 
-def generate_sepp_placements(seqs):
+def generate_sepp_placements(seqs, out_dir, threads, reference_phylogeny=None,
+                             reference_alignment=None):
     """Generates the sepp commands
 
     Parameters
     ----------
     seqs : list of str
         A list of seqs to generate placements
+    out_dir : str
+        The job output directory
+    threads : int
+        Number if CPU cores to use
+    reference_phylogeny : str, optional
+        A filepath to an alternative reference phylogeny for SEPP.
+        If None, default phylogeny is uses, which is Greengenes 13.8 99% id.
+    reference_alignment : str, optional
+        A filepath to an alternative reference alignment for SEPP.
+        If None, default alignment is uses, which is Greengenes 13.8 99% id.
 
     Returns
     -------
     dict of strings
         keys are the seqs, values are the new placements as JSON strings
     """
-    return {}
+    # return an empty dict if no sequences have been passed to the function
+    if len(seqs) < 1:
+        return {}
+
+    # Create a multiple fasta file for all input seqs
+    file_input = "%s/input.fasta" % out_dir
+    with open(file_input, 'w') as fh_input:
+        for seq in seqs:
+            fh_input.write(">%s\n%s\n" % (seq, seq))
+
+    # execute SEPP
+    run_name = 'qiita'
+    param_phylogeny = ''
+    if reference_phylogeny is not None:
+        param_phylogeny = ' -t %s ' % abspath(reference_phylogeny)
+    param_alignment = ''
+    if reference_alignment is not None:
+        param_alignment = ' -a %s ' % abspath(reference_alignment)
+    # SEPP writes output into the current working directory (cwd), therefore
+    # we here first need to store the cwd, then move into the output directory,
+    # perform SEPP and move back to the stored cwd for a clean state
+    curr_pwd = environ['PWD']
+    std_out, std_err, return_value = system_call(
+        'cd %s && run-sepp.sh %s %s -x %i %s %s; cd %s' %
+        (out_dir, file_input, run_name, threads,
+         param_phylogeny, param_alignment, curr_pwd))
+
+    # parse placements from SEPP results
+    try:
+        with open('%s/%s_placement.json' % (out_dir, run_name),
+                  'r') as fh_placements:
+            plcmnts = json.loads(fh_placements.read())
+            return {p['nm'][0][0]: p['p'] for p in plcmnts['placements']}
+    except OSError:
+        # due to the wrapper style of run-sepp.sh the actual exit code is never
+        # returned and we have no way of finding out which sub-command failed
+        # Therefore, we can only assume that something went wrong by not
+        # observing the expected output file.
+        return False, None, "Something went wrong with run-sepp.sh"
```

Comment:
Perhaps worth adding the exit number or the error? In fact, perhaps it should add an error but not stop the processing, right? I mean the deblur file was created the problem is assigning the placements.

Reply:
As explained further up, the exit code of the bash wrapper will always be 0 also for failing jobs.

You are right, the deblur process would be successful at this stage and only computation for unknown placements fail. So we can either say that the whole job fails (because a part failed) or we allow having deblur tables in qiita for which some features don't have placements.

I prefer the first design.